# Load the circlize package
library(circlize)
library(tidyverse)
library(ComplexHeatmap)
library(dplyr)

# Load the TE annotation GFF3 file
gff_file <- "hifi_assembly.fa.mod.EDTA.TEanno.gff3"
gff_data <- read.table(gff_file, header = FALSE, sep = "\t", stringsAsFactors = FALSE)

# Check the superfamilies present in the GFF3 file, and their counts
gff_data$V3 %>% table()


# custom ideogram data
## To make the ideogram data, you need to know the lengths of the scaffolds.
## There is an index file that has the lengths of the scaffolds, the `.fai` file (generated by samtools).
## Read this file in R and prepare the custom ideogram data

custom_ideogram <- read.table("hifi_assembly.fa.fai", header = FALSE, stringsAsFactors = FALSE)
custom_ideogram$chr <- custom_ideogram$V1
custom_ideogram$start <- 1
custom_ideogram$end <- custom_ideogram$V2
custom_ideogram <- custom_ideogram[, c("chr", "start", "end")]
custom_ideogram <- custom_ideogram[order(custom_ideogram$end, decreasing = T), ]
sum(custom_ideogram$end[1:20])

# Select only the first 18 longest scaffolds
custom_ideogram <- custom_ideogram[1:18, ]

# Function to filter GFF3 data based on Superfamily (You need one track per Superfamily)
filter_superfamily <- function(gff_data, superfamily, custom_ideogram) {
  filtered_data <- gff_data[gff_data$V3 == superfamily, ] %>%
    as.data.frame() %>%
    mutate(chrom = V1, start = V4, end = V5, strand = V6) %>%
    select(chrom, start, end, strand) %>%
    filter(chrom %in% custom_ideogram$chr)
  return(filtered_data)
}

# -------- Gene Track --------
gene_gff <- "filtered.genes.renamed.gff3"

genes_raw <- read.delim(gene_gff, header = FALSE, sep = "\t",
                        quote = "", comment.char = "#", stringsAsFactors = FALSE)
colnames(genes_raw)[1:9] <- c("seqid","source","type","start","end","score","strand","phase","attr")

# Filter for genes
genes <- genes_raw %>%
  dplyr::filter(type %in% c("gene")) %>%
  dplyr::transmute(chrom = seqid,
                   start = as.integer(start),
                   end   = as.integer(end)) %>%
  dplyr::filter(chrom %in% custom_ideogram$chr)

# -------- Plot Circles ------------
pdf("plots/TE_density_superfamilies.pdf", width = 11.5, height = 11.5)
gaps <- c(rep(1, length(custom_ideogram$chr) - 1), 5) # Add a gap between scaffolds, more gap for the last scaffold
circos.par(start.degree = 90, gap.after = 1, track.margin = c(0, 0), gap.degree = gaps)
# Initialize the circos plot with the custom ideogram
circos.genomicInitialize(custom_ideogram)

# Plot the density
circos.genomicDensity(filter_superfamily(gff_data, "Gypsy_LTR_retrotransposon", custom_ideogram), count_by = "number", col = "firebrick3", track.height = 0.07, window.size = 1e5)
circos.genomicDensity(filter_superfamily(gff_data, "Copia_LTR_retrotransposon", custom_ideogram), count_by = "number", col = "dodgerblue3", track.height = 0.07, window.size = 1e5)
circos.genomicDensity(filter_superfamily(gff_data, "CACTA_TIR_transposon", custom_ideogram), count_by = "number", col = "forestgreen", track.height = 0.07, window.size = 1e5)
circos.genomicDensity(filter_superfamily(gff_data, "Mutator_TIR_transposon", custom_ideogram), count_by = "number", col = "deeppink2", track.height = 0.07, window.size = 1e5)
circos.genomicDensity(filter_superfamily(gff_data, "hAT_TIR_transposon", custom_ideogram), count_by = "number", col = "goldenrod2", track.height = 0.07, window.size = 1e5)
circos.genomicDensity(filter_superfamily(gff_data, "tRNA_SINE_retrotransposon", custom_ideogram), count_by = "number", col = "darkturquoise", track.height = 0.07, window.size = 1e5)
circos.genomicDensity(filter_superfamily(gff_data, "L1_LINE_retrotransposon", custom_ideogram), count_by = "number", col = "darkviolet", track.height = 0.07, window.size = 1e5)
# Genes
circos.genomicDensity(genes, count_by = "number", col = "grey35", track.height = 0.07, window.size = 1e5)
circos.clear()

lgd_te <- Legend(
  title = "Superfamily", at = c("Gypsy LTR retrotransposon", "Copia LTR retrotransposon", "CACTA TIR transposon", "Mutator TIR transposon", "hAT TIR transposon", "tRNA SINE retrotransposon", "L1 LINE retrotransposon"),
  legend_gp = gpar(fill = c("firebrick3", "dodgerblue3", "forestgreen", "deeppink2", "goldenrod2", "darkturquoise", "darkviolet")))

lgd_gene <- Legend(
  title = "Gene Track", at = "Gene density",
  legend_gp = gpar(fill = "grey35"))

lgd_all <- packLegend(lgd_te, lgd_gene, direction = "vertical", gap = unit(4, "mm"))

draw(lgd_all, x = unit(4, "cm"), y = unit(26, "cm"), just = c("topleft"))

dev.off()

# ---------------------------------------------------------------------------------------
# Plot the distribution of Athila and CRM clades (known centromeric TEs in Brassicaceae).

gypsy_cls <- read.delim(file = "Gypsy_sequences.fa.rexdb-plant.cls.tsv", sep = "\t", skip=1, header=FALSE)
copia_cls <- read.delim(file = "Copia_sequences.fa.rexdb-plant.cls.tsv", sep = "\t", skip=1, header=FALSE)

# Column names
column_names <- c("TE", "Order", "Superfamily", "Clade", "Complete", "Strand", "Domains")
names(gypsy_cls) <- column_names
names(copia_cls) <- column_names

# Extract the TE ID
gypsy_cls$id <- str_extract(gypsy_cls$TE, "TE_[0-9]+")
copia_cls$id <- str_extract(copia_cls$TE, "TE_[0-9]+")
# from EDTA gff3 file
gff_data$id <- stringr::str_extract(gff_data$V9, "TE_[0-9]+")

# Join tables
gff_gypsy <- inner_join(gff_data, gypsy_cls, by = "id")

# Get CRM and Athila
crm <- gff_gypsy %>% filter(Clade == "CRM")
athila <- gff_gypsy %>% filter(Clade == "Athila")

crm_circos <- crm %>%
  mutate(
    chrom = V1,
    start = pmin(V4, V5),
    end   = pmax(V4, V5)
  ) %>%
  select(chrom, start, end) %>%
  filter(chrom %in% custom_ideogram$chr)

athila_circos <- athila %>%
  mutate(
    chrom = V1,
    start = pmin(V4, V5),
    end   = pmax(V4, V5)
  ) %>%
  select(chrom, start, end) %>%
  filter(chrom %in% custom_ideogram$chr)

pdf("plots/TE_density_gypsy_clades.pdf", width = 11.5, height = 11.5)
gaps <- c(rep(1, length(custom_ideogram$chr) - 1), 5)

circos.clear()
circos.par(start.degree = 90, gap.after = gaps, track.margin = c(0, 0), gap.degree = gaps)
circos.genomicInitialize(custom_ideogram)

# All Gypsy
circos.genomicDensity(filter_superfamily(gff_data, "Gypsy_LTR_retrotransposon", custom_ideogram), count_by = "number", col = "firebrick3", track.height = 0.1, window.size = 1e5)

# CRM
circos.genomicDensity(crm_circos, count_by = "number", col = "purple4", track.height = 0.1, window.size = 1e5)

# Athila
circos.genomicDensity(athila_circos, count_by = "number", col = "pink3", track.height = 0.1, window.size = 1e5)

circos.clear()

lgd_gypsy <- Legend(
  title = "Superfamily", at = c("Gypsy LTR retrotransposon"),
  legend_gp = gpar(fill = c("firebrick3")))
lgd_clades <- Legend(
  title = "Gypsy clades",
  at = c("CRM", "Athila"),
  legend_gp = gpar(fill = c("purple4", "pink3")))
lgd_all <- packLegend(lgd_gypsy, lgd_clades, direction = "vertical", gap = unit(4, "mm"))
draw(lgd_all, x = unit(4, "cm"), y = unit(26, "cm"), just = c("topleft"))

dev.off()

## --------- Now plot all tracks together --------- ##
pdf("plots/TE_density_all.pdf", width = 12, height = 11.5)
gaps <- c(rep(1, length(custom_ideogram$chr) - 1), 5)
circos.clear()
circos.par(start.degree = 90, gap.after = gaps, track.margin = c(0, 0), gap.degree = gaps)
circos.genomicInitialize(custom_ideogram)

# Plot the densities
# Genes
circos.genomicDensity(genes, count_by = "number", col = "grey25", track.height = 0.07, window.size = 1e5)
# Superfamilies
circos.genomicDensity(filter_superfamily(gff_data, "Gypsy_LTR_retrotransposon", custom_ideogram), count_by = "number", col = "red3", track.height = 0.07, window.size = 1e5)
# Gypsy clades (CRM & Athila)
circos.genomicDensity(crm_circos, count_by = "number", col = "purple4", track.height = 0.07, window.size = 1e5)
circos.genomicDensity(athila_circos, count_by = "number", col = "pink3", track.height = 0.07, window.size = 1e5)
# More Superfamilies
circos.genomicDensity(filter_superfamily(gff_data, "Copia_LTR_retrotransposon", custom_ideogram), count_by = "number", col = "dodgerblue3", track.height = 0.07, window.size = 1e5)
circos.genomicDensity(filter_superfamily(gff_data, "CACTA_TIR_transposon", custom_ideogram), count_by = "number", col = "forestgreen", track.height = 0.07, window.size = 1e5)
circos.genomicDensity(filter_superfamily(gff_data, "Mutator_TIR_transposon", custom_ideogram), count_by = "number", col = "deeppink3", track.height = 0.07, window.size = 1e5)
circos.genomicDensity(filter_superfamily(gff_data, "hAT_TIR_transposon", custom_ideogram), count_by = "number", col = "goldenrod2", track.height = 0.07, window.size = 1e5)
circos.genomicDensity(filter_superfamily(gff_data, "tRNA_SINE_retrotransposon", custom_ideogram), count_by = "number", col = "darkturquoise", track.height = 0.07, window.size = 1e5)
circos.genomicDensity(filter_superfamily(gff_data, "L1_LINE_retrotransposon", custom_ideogram), count_by = "number", col = "violet", track.height = 0.07, window.size = 1e5)

circos.clear()
# Add Legend
lgd_te <- Legend(
  title = "Superfamily", at = c("Gypsy LTR retrotransposon", "Copia LTR retrotransposon", "CACTA TIR transposon", "Mutator TIR transposon", "hAT TIR transposon", "tRNA SINE retrotransposon", "L1 LINE retrotransposon"),
  legend_gp = gpar(fill = c("red3", "dodgerblue3", "forestgreen", "deeppink2", "goldenrod2", "darkturquoise", "violet")))
lgd_gene <- Legend(
  title = "Gene Track", at = "Gene density",
  legend_gp = gpar(fill = "grey35"))
lgd_clades <- Legend(
  title = "Gypsy clades",
  at = c("CRM", "Athila"),
  legend_gp = gpar(fill = c("purple4", "pink3")))

lgd_te_gene <- packLegend(lgd_te, lgd_gene, direction = "vertical", gap = unit(4, "mm"))
lgd_all <- packLegend(lgd_te_gene, lgd_clades, direction = "horizontal", gap = unit(7, "mm"))
draw(lgd_all, x = unit(6, "cm"), y = unit(26, "cm"), just = c("topleft"))

dev.off()